qiskit_ibm_runtime/transpiler/passes/basis/fold_rzz_angle.py

```diff
@@ -0,0 +1,185 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Pass to wrap Rzz gate angle in calibrated range of 0-pi/2."""
+
+from typing import Tuple
+from math import pi
+
+from qiskit.circuit.library.standard_gates import RZZGate, RZGate, XGate
+from qiskit.circuit.parameterexpression import ParameterExpression
+from qiskit.circuit import Qubit
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.transpiler.basepasses import TransformationPass
+
+import numpy as np
+
+
+class FoldRzzAngle(TransformationPass):
+    """Fold Rzz gate angle into calibrated range of 0-pi/2 with
+    local gate tweaks.
+
+    This pass preserves the number of Rzz gates, but it may add
+    extra single qubit gate operations.
+    These gates might be reduced by the following
+    single qubit optimization passes.
+    """
+
+    def run(self, dag: DAGCircuit) -> DAGCircuit:
+        # Currently control flow ops and Rzz cannot live in the same circuit.
+        # Once it's supported, we need to recursively check subroutines.
+        for node in dag.op_nodes():
+            if not isinstance(node.op, RZZGate) or isinstance(
+                angle := node.op.params[0], ParameterExpression
+            ):
+                continue
+            wrap_angle = np.angle(np.exp(1j * angle))
+            if -pi <= wrap_angle <= -pi / 2:
+                replace = _quad3(wrap_angle, node.qargs)
+            elif -pi / 2 < wrap_angle < 0:
+                replace = _quad4(wrap_angle, node.qargs)
+            elif pi / 2 < wrap_angle <= pi:
+                replace = _quad2(wrap_angle, node.qargs)
+            else:
+                if angle != wrap_angle:
+                    dag.substitute_node(
+                        node,
+                        RZZGate(wrap_angle),
+                        inplace=True,
+                    )
+                continue
```

Comment:
Please explain, thanks

Reply:
I understand what confuses me. On the one hand, from what I  see the range to consider should be [-2pi, 2pi] (see https://github.com/Qiskit/qiskit-ibm-runtime/pull/2043/files#r1841666155). On the other hand, I must get something wrong, because according to the code here the period is 2pi.

Reply:
After phase wrapping you always get [-pi, pi]